middle_end/flambda2.0/from_lambda/lambda_to_flambda_primitives.ml

```diff
@@ -255,38 +255,46 @@ let convert_lprim ~backend (prim : L.primitive) (args : Simple.t list)
   let module B = (val backend : Flambda2_backend_intf.S) in
   let size_int = B.size_int in
   match prim, args with
-  | Pmakeblock (tag, flag, shape), _ ->
-    let flag = C.convert_mutable_flag flag in
+  | Pmakeblock (tag, mutability, shape), _ ->
+    let tag = Tag.Scannable.create_exn tag in
     let shape = C.convert_block_shape shape ~num_fields:(List.length args) in
-    Variadic (Make_block (
-        Full_of_values (Tag.Scannable.create_exn tag, shape), flag),
-      args)
-  | Pmakearray (kind, mutability), _ ->
-    let flag = C.convert_mutable_flag mutability in
-    let kind, args =
-      let module S = P.Generic_array_specialisation in
-      match kind with
-      | Pgenarray -> S.no_specialisation (), args
-      | Paddrarray -> S.full_of_arbitrary_values_but_not_floats (), args
-      | Pintarray -> S.full_of_immediates (), args
-      | Pfloatarray -> S.full_of_naked_floats (), List.map unbox_float args
+    let mutability = C.convert_mutable_flag mutability in
+    Variadic (Make_block (Values (tag, shape), mutability), args)
+  | Pmakearray (array_kind, mutability), _ ->
+    let array_kind = C.convert_array_kind array_kind in
+    let mutability = C.convert_mutable_flag mutability in
+    let args =
+      match array_kind with
+      | Float_array_opt_dynamic | Immediates | Values -> args
+      | Naked_floats -> List.map unbox_float args
     in
-    Variadic (Make_block (Generic_array kind, flag), args)
+    Variadic (Make_array (array_kind, mutability), args)
   | Popaque, [arg] ->
     Unary (Opaque_identity, arg)
   | Pduprecord (repr, num_fields), [arg] ->
-    let kind : P.duplicate_block_kind =
+    let kind : P.Duplicate_block_kind.t =
       match repr with
-      | Record_regular -> Full_of_values_known_length Tag.Scannable.zero
+      | Record_regular ->
+        Values {
+          tag = Tag.Scannable.zero;
+          length = Targetint.OCaml.of_int num_fields;
+        }
       | Record_float ->
-        Full_of_naked_floats
-          { length = Some (Targetint.OCaml.of_int num_fields) }
+        Naked_floats {
+          length = Targetint.OCaml.of_int num_fields;
+        }
       | Record_unboxed _ ->
         Misc.fatal_error "Pduprecord of unboxed record"
       | Record_inlined tag ->
-        Full_of_values_known_length (Tag.Scannable.create_exn tag)
+        Values {
+          tag = Tag.Scannable.create_exn tag;
+          length = Targetint.OCaml.of_int num_fields;
+        }
       | Record_extension _ ->
-        Full_of_values_known_length Tag.Scannable.zero
+        Values {
+          tag = Tag.Scannable.zero;
+          length = Targetint.OCaml.of_int num_fields;
```

Comment:
This should be `num_fields + 1` here, as the fields do not contain the extra field for the hashed constructor.

Reply:
Well spotted, ugh.